### DISCUSSION TITLE
max_active_runs_per_dag is not respected

### DISCUSSION THREAD

**Discussion by HenryLee19** [apache/airflow#38165] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.8.3

### What happened?

I changed  the "max_active_runs_per_dag" from the default of 16 to 2, in the Airflow config. Then manually triggered a DAG 10x via the UI and observed 10 instances of the DAG in the "running" state.

### What you think should happen instead?

I expected there to be 2 instances in the "running" state and 8 in the "queued" state. 

(Note, setting "max_active_runs" at the DAG level works. It is only the global "max_active_runs_per_dag" that appears to be not working.)

### How to reproduce

1. in airflow.cfg, change "max_active_runs_per_dag" from the default of 16 to 2
2. restart Airflow (Docker container, in my case)
3. trigger a DAG via the UI play button 10x
4. observe 10 instances of the DAG in the "running" state

### Operating System

Debian GNU/Linux 12 (bookworm)

### Deployment

Docker-Compose

### Deployment details

Used the docker-compose found here: https://airflow.apache.org/docs/apache-airflow/2.8.3/docker-compose.yaml

### Anything else?

DAG definition:

```
from datetime import datetime
import time

from airflow import DAG
from airflow.operators.python import PythonOperator

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'start_date': datetime(2024, 1, 1),
    'email': ['airflow@example.com'],
    'email_on_failure': False,
    'email_on_retry': False,
    'retries': 0,
}


def _task1(**context):
    time.sleep(120)
    return


with DAG(
    dag_id='dag1',
    default_args=default_args,
    schedule=None,
    catchup=False
) as dag:

    task1 = PythonOperator(
        task_id='task1',
        python_callable=_task1
    )
```

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by jscheffl** [comment]
Thanks for reporting the bug. I very much doubt that in 2.8.3 this behavior is changed. It is a core feature in Scheduler to restrict runs if set by parameter.

My first guess is that the scheduler does not effectively have the right parameter being set. Can you please:
1. Copy the finally used airflow config here
2. Attach the final used compose file to the bug
3. Exec into the scheduler POD and execute the `env` command validating the the config is present. Maybe also post the output here.

**Reply by HenryLee19** [reply]
Thanks for the quick reply. Here is the airflow.cfg located at /opt/airflow in my Docker Airflow webserver. (Extension changed to "txt" to allow upload.)
[airflow.cfg.txt](https://github.com/apache/airflow/files/14608551/airflow.cfg.txt)

**Reply by HenryLee19** [reply]
Also here is the docker-compose that was used. (Extension changed to "txt" to allow upload.)
[docker-compose.yaml.txt](https://github.com/apache/airflow/files/14608560/docker-compose.yaml.txt)

**Reply by HenryLee19** [reply]
Regarding, the `env` command mentioned above...indeed, I do not see that environment variable inside the containers. However, the airflow.cfg has over 350 environment variables and I don't see most of them.

I also checked for this environment variable inside our GCP Composer environment and do not see it there either. I checked by trying to log the value from inside a DAG: `logging.info(os.getenv('AIRFLOW__CORE__MAX_ACTIVE_RUNS_PER_DAG', 'foo'))`

So, is there something else that needs to be done to enable the variables that are in the "airflow.cfg"?

**Reply by potiuk** [reply]
You can login to the system and run `airlfow info`- this will show you the actual values used by `airflow` CLI  (providing that you run it in exact same environment as scheduler)

**Reply by potiuk** [reply]
Actually `airflow config list -s`  -> this one will dump you the configuraiton including expliaining where airflow read the configuration from.

**Reply by potiuk** [reply]
And yes - you won't see variables configured in .cfg as env vars. It's impossible if you know how env vars work.

**Reply by HenryLee19** [reply]
Ok, then perhaps I misunderstood your 3rd bullet point above regarding `env`. I thought you were saying I should see environment variable `AIRFLOW__CORE__MAX_ACTIVE_RUNS_PER_DAG`.

**Reply by HenryLee19** [reply]
`airflow config list -s` shows `max_active_runs_per_dag = 2`

**Reply by potiuk** [reply]
Because the assumption was that you do use env variables,

Please do more experimenting and digging and post more evidences - gits wiht outputs of your commands and list your pods and output of your termial showing the way how you check it - explining whch pod is which and how you get there. 

As you see - we cannot reproduce it - so we have the reason to believe your settnig is not applied to running scheduler. Probably you made some mistake and the confiiguration you set has not been applied - one most common mistake people do is they do not restart components - after changing the configuration they do not restart the components, or they do not check the valuies in the pod that is actually used to run the component, or their variables are set differently because their components are running in non-interactive shell lhttps://medium.com/@linuxschooltech/what-are-interactive-and-non-interactive-shells-in-linux-5f25ce19e537 and gets different variables and even different configuration from different place.

**Reply by HenryLee19** [reply]
Indeed, this appears to be an issue of the configuration not being applied. 

I proved that 2 ways: First, I did not attempt to change the config and simply left the default of 16 and started a lot more than that. As originally expected, only 16 ran at a given time, the rest were queued. Second, setting the environment variable equivalent `AIRFLOW__CORE__MAX_ACTIVE_RUNS_PER_DAG` behaved similarly, as expected.

For my own curiosity, I'd love to know how to get this working via the `airflow.cfg` in the Docker container. As you mentioned, and as I attempted by restarting the containers, I imagine it has something to do with restarting the various services. But inside the container, the airflow user does not have such permissions and I am out of time to take this further. 

At the moment, knowing I can test the various configurations by setting environment variables is good enough, even if I can't get the `airflow.cfg` to work as I expected.

**Comment by potiuk** [comment]
Yep. Converting to discussion - as it seems problem with configuring it . We can always convert back to issue if there are evidences and reproducibility.

**Comment by ephraimbuddy** [comment]
I reproduced this in the current main and will investigate more later. Not sure when this broke but it seems broken

**Reply by potiuk** [reply]
Created the issue https://github.com/apache/airflow/issues/38167 (back) out of it

**Reply by ephraimbuddy** [reply]
Turned out there was a mistake in the setting for the environment variable I set. This is not reproducible in the current main. I will check later with released versions
